### PR TITLE
New project page layout and JSON API

### LINKED ISSUE #16: Provide an API
It would be nice if we could provide some information via an API - nothing fancy, just some static JSON generated from the DB.

Either one big JSON blob containing everything, or perhaps a few more distinct sets of data.

I'm thinking:

- List of packages (could be limited to those we have available, or contain metadata including success / build output
- List of versions per package
- Build output

### LINKED ISSUE #157: Show dependencies per file
We should be able to show the relevant apt dependencies for each file, not just the latest.

### LINKED ISSUE #178: Show "all" as ABI in versions table for pure python wheels
In the versions table on project pages, instead of just listing the lowest ABI (cp34m), it would be better if we were able to deduce that it's pure python and show "all".

Screenshot from 2019-07-02 13-34-37

It may give the impression that we only have a working wheel for Jessie.

### LINKED ISSUE #241: New project page layout
Some ideas of ways to better present the contents of the project pages:

Screenshot from 2020-08-12 00-38-00

It might be nice to nest the files within each version row, rather than list them in a separate table. Perhaps they could be expandable with a button:

Screenshot from 2020-08-12 01-03-16

### PR DESCRIPTION
**Pull request by bennuttall**
- Closes #241
- Closes #157
- Closes #178
- Closes #16 

TODO:

- [x] Add tests
- [x] Update db upgrade script
- [x] Add code for deletion of `/project/<pkg>/json/` and its index file

### CODE REVIEW

**Comment by bennuttall** [comment]
@waveform80 this is complete now - would you mind giving it a review?

One thing I'd like to change is to move the file apt dependencies lookup into the project files query. I previously solved this in the previous json branch [using a join and string_agg](https://github.com/piwheels/piwheels/blob/json/piwheels/initdb/sql/create_piwheels.sql#L995:L999) but now we've introduced the [get_file_apt_dependencies](https://github.com/piwheels/piwheels/blob/241-project-page/piwheels/initdb/sql/create_piwheels.sql#L1513) function we need a way of aggregating the result of that function call into a single row OR moving the logic of that function into the project files query. What do you think?

**Comment by bennuttall** [comment]
Oh the tests are failing on Travis because I'm trying to read the json file after writing. Doesn't look any different from the `packages.json` test though :/

**Comment by bennuttall** [comment]
@waveform80 this is now ready for review

File: piwheels/master/the_scribe.py [hunk 1]
```diff
@@ -52,6 +53,35 @@
...
+UTC = timezone.utc
+
+
+ProjectRelease = namedtuple('ProjectRelease', (
+    'version', 'yanked', 'released', 'skip', 'stretch', 'buster', 'files',
+    'builds_succeeded', 'builds_failed'))
+
+def _pure_python_from_lower_abi(abi, files):
+    return any(
+        f['file_abi_tag'] == 'none' and abi > f['builder_abi']
```
Comment (waveform80) [inline piwheels/master/the_scribe.py:65]: Is this meant to be abi greater-than or abi greater-than-or-equal-to? The latter would *seem* to be what we want but maybe I'm missing something.

File: piwheels/master/the_scribe.py [hunk 2]
```diff
@@ -263,26 +294,50 @@ def write_simple_index(self):
             index.file.write(self.templates['simple_index'](
                 packages=self.package_cache))
 
-    def write_package_index(self, package, *, exclude=None):
+    def write_pages(self, package, both=False, exclude=None):
```
Comment (waveform80) [inline piwheels/master/the_scribe.py:297]: What happened to my keywords-only "*"? :)

File: piwheels/master/the_scribe.py [hunk 3]
```diff
@@ -355,6 +399,56 @@ def write_project_page(self, package, *, exclude=None):
...
+                }
+            }
+            releases_dict[release.version.original] = release_dict
+        project_data = {
+            'package': package,
+            'summary': description,
+            'updated': datetime.now(tz=UTC).strftime('%Y-%m-%d %H:%M:%S %Z'),
+            'pypi_url': 'https://pypi.org/project/{}'.format(package),
+            'piwheels_url': 'https://www.piwheels.org/project/{}'.format(package),
+            'releases': releases_dict,
```
Comment (waveform80) [inline piwheels/master/the_scribe.py:445]: Oh, I would've be soooo tempted to dict-comprehension releases_dict ;)

File: piwheels/master/the_scribe.py [hunk 4]
```diff
@@ -378,26 +472,31 @@ def delete_package(self, package):
 
         pkg_dir = self.output_path / 'simple' / package
         proj_dir = self.output_path / 'project' / package
+        proj_json_dir = proj_dir / 'json'
+        canon_pkg_dir = pkg_dir.with_name(canonicalize_name(pkg_dir.name))
+        canon_proj_dir = pkg_dir.with_name(canonicalize_name(pkg_dir.name))
```
Comment (waveform80) [inline piwheels/master/the_scribe.py:477]: These are redundant aren't they?

File: piwheels/master/the_scribe.py [hunk 5]
```diff
@@ -420,6 +519,91 @@ def delete_version(self, package, version):
...
+                version=version.version,
+                yanked=version.yanked,
+                released=version.released,
+                skip=version.skip,
+                stretch=_get_abi_status(
+                    'cp35m', builds_succeeded, builds_failed, version_files,
+                    version.skip),
+                buster=_get_abi_status(
+                    'cp37m', builds_succeeded, builds_failed, version_files,
+                    version.skip),
```
Comment (waveform80) [inline piwheels/master/the_scribe.py:599]: Urggggggh - hard-coded ABIs *and* release names. Surely there's a better way of doing this? I wouldn't mind this sort of stuff in the *templates* as ultimately they probably want to be configurable for an installation anyway, but having this hard-coded in the master itself .... urgh.

**Review by waveform80 — CHANGES_REQUESTED** [review]
Everything's minor *but* for that one bit in the scribe with explicit ABIs & code-names

**Comment by bennuttall** [comment]
[](https://twitter.com/overflow_meme/status/1307512035278811138)

https://twitter.com/overflow_meme/status/1307512035278811138

**Review by waveform80 — APPROVED** [review]

**Pull request merged**